### PR TITLE
Dq Pluggability

### PR DESCRIPTION
**Pull request by arn7374**
# DQ_PLUGGABILITY
## Summary
This change has added the ability for users to plug in theirown DQ functionality through a custom plugin of your their construction. This means that existing DQ functionality has been moved to a plugin which is enabled by default, and which can be overwritten by their own.

## Areas Affected
*Place an 'x' within the braces to check the box*
- [X ] MAMAC
- [X] MAMACPP
- [ ] MAMADOTNET
- [ ] MAMAJNI
- [ ] MAMDA
- [ ] MAMDACPP
- [ ] MAMDADOTNET
- [ ] MAMDAJNI
- [ ] Visual Studio
- [X] SCons
- [X] Unit Tests
- [ ] Examples

## Details
This PR contains work that means default functionality can be used, overwritten with the user’s own functionality, or disabled altogether. This functionality can be configured on a per transport basis, so each transport can have its own data quality functions to allow transports consuming data from different sources to execute different DQ logic for each source. Users will be able to provide their own functionality by providing a shared object and referring to it in a properties file.

## Testing
I have tested this change with a selection of new Unit Tests
As well as this the functionality has been tested via the following means:
- verified hooks can be seen in the logging but no data is received with mama.wmw.dq_enabled=1 and custom plugin is defined in mama.properties
- Verified hooks can be seen in the logging and data is received with mama.wmw.d_enabled=0 and plugin is defined in mama.properties
- Verified hooks cannot be seen in the logging but data can be received with mama.wmw.dq_enabled=1 and custom plugin is not defined in mama.properties
- Verified hooks cannot be seen in the logging but data can be received with mama.wmw.dq_enabled=0 and custom plugin is not defined in mama.properties
- verified hooks can be seen in the logging but no data is received with mama.wmw.dq_enabled not specified in mama.properties (default value "enabled" used) and custom plugin is defined in mama.properties
- Verified hooks cannot be seen in the logging but data can be received with mama.wmw.dq_enabled(mama.wmw.dq_enabled not specified in mama.properties (default value "enabled" used) and custom plugin is not defined in mama.properties
- Verified specifying multiple custom plugins
- Verified specifying custom plugin per transport
- Verified plugins successfully loaded
- Verified all hooks defined in the custom plugin successfully loaded
- Verified with Market Data
- Verified with Initilas, Updates, Quotes and Trades
- Verified the inability of using a custom plugin when so file not available in the lib folder but the plugin defined in mama.properties
- Verified the inability of using a custom plugin when wrong name provided in mama.properties
- Verified with single/multiples clients 
- verified with churning/restarting pub/sub
- Verified with missing optional hooks in custom plugin
- Verified with missing required hooks in custom plugin
- Verified with multiple plugins on the same transport
- Verified with valid/invalid values for dq_enabled
- verified with default value for dq_enabled
- Verified other functionality not affected
- Verified on Linux and Windows

### CODE REVIEW

File: mama/c_cpp/src/c/subscription.c [hunk 1]
```diff
@@ -658,8 +655,7 @@ mamaSubscription_setupBasic (
     }
 
     /*Create the DQ strategy - sequence number checking etc.*/
-    if (MAMA_STATUS_OK!=(status=dqStrategy_create (&self->mDqStrategy,
-                      subscription)))
+    if (MAMA_STATUS_OK!=(status=mamaPlugin_fireSubscriptionPostCreateHook (subscription)))
     {
         mama_log (MAMA_LOG_LEVEL_ERROR,
                   "Could not create DQ strategy. [%s]",
```
Comment (fquinner) [inline mama/c_cpp/src/c/subscription.c]: This section is no longer responsible for creating dq so comment and logging should reflect / be moved. Also conscious that plugin only looks to preserve the *last* status code so if multiple hooks are registered, this error detection will not work.

Comment (arn7374) [reply mama/c_cpp/src/c/subscription.c]: good call, that logging is in the hook in DQ strategy plugin, the plugin.c code handles the error code handling

File: mama/c_cpp/src/c/dqstrategyplugin/dqstrategyplugin.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
+/*
+ * mamaPluginTemplate.h
+ *
+ *  Created on: 23 Mar 2015
+ */
+
+#ifndef MAMAPLUGINHOOK_H_
```
Comment (dmagOM) [inline mama/c_cpp/src/c/dqstrategyplugin/dqstrategyplugin.h]: Header guards are duplicated from the plugin hook example, which is going to cause issues with the build. Should be fixed.

File: mama/c_cpp/src/c/dqstrategyplugin/dqstrategyplugin.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
+/*
+ * mamaPluginTemplate.h
```
Comment (dmagOM) [inline mama/c_cpp/src/c/dqstrategyplugin/dqstrategyplugin.h]: These comments should be tidied up, they make no sense in these files.

File: mama/c_cpp/src/c/dqstrategyplugin/dqstrategyplugin.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+mama_status
+dqstrategyMamaPlugin_transportPostCreateHook (mamaPluginInfo pluginInfo, mamaTransport transport);
+
+PLUGINExpDLL
+mama_status
+dqstrategyMamaPlugin_publisherPreSendHook (mamaPluginInfo pluginInfo, mamaPublisher publisher, mamaMsg message);
+
+PLUGINExpDLL
+mama_status
+dqstrategyMamaPlugin_transportEventHook (mamaPluginInfo pluginInfo, mamaTransport transport, int setStale, mamaTransportEvent tportEvent);
```
Comment (dmagOM) [inline mama/c_cpp/src/c/dqstrategyplugin/dqstrategyplugin.h]: A lot of these lines are pretty significantly beyond the 80 character mark, which makes them harder to read. You should be able to run these files through the clang-format rules and get something that matches the style guide, but for reference it's discussed here: https://openmama.github.io/openmama_coding_standards.html#line-width

**Comment by fquinner** [comment]
Just a reminder that the windows build is still failing:

http://ci.openmama.org:8080/job/arn7374-DQ_PLUGGABILITY-windows/lastFailedBuild/console

**Comment by fquinner** [comment]
Compiles on windows now :)

Introduces a new compiler warning though :(

http://ci.openmama.org:8080/job/arn7374-DQ_PLUGGABILITY-windows/lastSuccessfulBuild/warnings40Result/new/

**Comment by arn7374** [comment]
Thanks, windows should be stable now.

**Comment by fquinner** [comment]
Tried building this with visual studio and it did work ok though it needed this change:

```
diff --git a/mama/c_cpp/src/c/generateMamaSourceFiles.bat b/mama/c_cpp/src/c/generateMamaSourceFiles.bat
index feee5a11..771027cf 100644
--- a/mama/c_cpp/src/c/generateMamaSourceFiles.bat
+++ b/mama/c_cpp/src/c/generateMamaSourceFiles.bat
@@ -1,7 +1,7 @@
 set BUILD_DIR=%1
 set VERSION_MAJOR=6
 set VERSION_MINOR=2
-set VERSION_RELEASE="2_rc2"
+set VERSION_RELEASE=1
 set RC_VERSION_NUMERICAL=1
 set RC_VERSION_STRING=1
 set MAMACDLL=libmamacmdd.dll
```

However it doesn't look like there are any visual studio project files existing for the dq strategy plugin so those trying to build from the solution files will be unable to build a working copy of openmama any more because they will get this error message when they try to run since the plugin didn't build, after which point the example applications will bravely try to persevere before falling over:

```
2018-03-14 21:24:36: (1878) : mama_loadPlugin(): Could not open plugin library [dqstrategy] [126]
```

**Comment by fquinner** [comment]
Any update here on fixing the project files?

**Comment by fquinner** [comment]
Thanks Aaron, as I mentioned offline, I have given you write access to the project. Can you put together a comment to summarize the change and squash commit it into `next` so that

1. The commit log is cleaned up and pruned
2. You can maintain ownership of the changes (for contact in git blame etc).

**Comment by fquinner** [comment]
6de60350f40b4d9b0f181b7e64e0d192798b9f97

**Pull request closed without merging**